### ISSUE TITLE
Vue 3 defineProps and defineEmits having issues

### ISSUE THREAD

**Issue by justgrumpy** [stryker-mutator/stryker-js#3305]
**Summary**

I am having issues using Stryker with the Vue 3 `<script setup>` for Composition API inside Single File Components (SFCs). Specifically I am having issues with `defineProps` and `defineEmits`, https://v3.vuejs.org/api/sfc-script-setup.html#defineprops-and-defineemits. I get an error like the following:

```
13:38:55 (13792) ERROR DryRunExecutor One or more tests resulted in an error:
        undefined[@vue/compiler-sfc] `defineProps()` in <script setup> cannot reference locally declared variables because it 
will be hoisted outside of the setup() function. If your component options require initialization in the module scope, use a separate normal <script> to export the options instead.

C:\dev\my-vue-app\stryker-tmp\sandbox309818\src\components\HelloWorld.vue
65 |
66 |  import { ref } from 'vue';
67 |  defineProps(stryMutAct_9fa48("0") ? {} : (stryCov_9fa48("0"), {
   |              ^^^^^^^^^^^^^^^^
68 |    msg: String
69 |  })); Error: [@vue/compiler-sfc] `defineProps()` in <script setup> cannot reference locally declared variables because it will be hoisted outside of the setup() function. If your component options require initialization in the module scope, use a 
separate normal <script> to export the options instead.
```

The app works normally and the Jest tests run fine. I assume that Stryker in mutating is doing something to reference locally declared variables causing this error???

A full sample application with the problem can be found here, https://github.com/justgrumpy/my-vue-app.

**Stryker config**

```json
{
    "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
    "mutator": {
        "plugins": []
    },
    "tempDirName_comment": "Jest doesn't play nice with the default (.stryker-tmp).",
    "tempDirName": "stryker-tmp",
    "testRunner": "jest",
    "coverageAnalysis": "off"
}
```

**Test runner config**

**Stryker environment**

```
+-- @stryker-mutator/core@5.5.1
+-- @stryker-mutator/jest-runner@5.5.1
```

```
+-- @stryker-mutator/jest-runner@5.5.1
+-- @vue/vue3-jest@27.0.0-alpha.4
+-- babel-jest@27.4.2
+--  jest@27.4.3
```

**Test runner environment**

```shell
jest
```

babel.config.js
```json
module.exports = {
  presets: [['@babel/preset-env', { targets: { node: 'current' } }]],
};
```

jest.config.js
```json
const {defaults} = require('jest-config');
module.exports = {
  moduleFileExtensions: [...defaults.moduleFileExtensions, 'vue'],
  transform: {
    '^.+\\.js$': 'babel-jest',
    '^.+\\.vue$': '@vue/vue3-jest',
  },
  testEnvironment: 'jsdom',
};
```

**Your Environment**

| software         | version(s)
| ---------------- | -------
| node             | v16.13.0
| npm              | 8.1.0
| Operating System |

**Add stryker.log**

[stryker.log](https://github.com/stryker-mutator/stryker-js/files/7678900/stryker.log)

**Comment by nicojs** [comment]
Hi @justgrumpy thanks for opening this issue.

> I assume that Stryker in mutating is doing something to reference locally declared variables causing this error???

Yes, it does. You can see it in your error message:

```
67 |  defineProps(stryMutAct_9fa48("0") ? {} : (stryCov_9fa48("0"), {
   |              ^^^^^^^^^^^^^^^^
68 |    msg: String
```

The variable `stryMutAct_9fa48` is inserted in the top of the `<script>` tag. You can also see that for yourself inside the `stryker-tmp` sandbox directory after a failing run.

The problem seems to be that [`defineProps` and `defineEmits`](https://v3.vuejs.org/api/sfc-script-setup.html#defineprops-and-defineemits) are **compiler macros** instead of JavaScript functions. 

The only good solution (I can think of) for these kinds of problems is a feature called "ignore-plugins" described in #3229. We could create a specific `vue-ignore`-plugin that automatically ignores compiler macros.

As a workaround, you can insert a [stryker disable comment](https://stryker-mutator.io/docs/stryker-js/disable-mutants#using-a--stryker-disable-comment)

```diff
<script setup>
import { ref } from 'vue'

+// Stryker disable next-line all: Don't mutate compiler macros
defineProps({
  msg: String
})

const count = ref(0)
</script>
```

Does that work for you?

**Comment by justgrumpy** [comment]
This does work for me. I was unaware of the ability to disable Stryker for a specific line of code. This appears to be a feature added with 5.4?

An ignore-plugin would also be very convenient to eliminate so many comments. If there are several components, comments will be needed in each `.vue` file. If more than one property is defined and object-based syntax is used to define the property (https://v3.vuejs.org/api/options-data.html#props), several comments are required in the same file...

```
// Stryker disable next-line all: Don't mutate compiler macros
defineProps({
  // Stryker disable next-line all: Don't mutate compiler macros
  msg: { type: String, required: true },
  // Stryker disable next-line all: Don't mutate compiler macros
  anotherProp: { type: String, required: false }
})
```

**Comment by nicojs** [comment]
Yes, it was added in 5.4. We're rapidly evolving currently, focusing on user experience and performance. Following [@stryker_mutator](https://twitter.com/stryker_mutator/) on Twitter helps with keeping up to date. Our [release notes](https://github.com/stryker-mutator/stryker-js/releases) should also be quite readable. About once a year, we also write a blog article with some bigger changes here: https://stryker-mutator.io/blog

I'm first focussing on improving support for native ESM (see #2922), after that I want to focus on ignore plugins. Let's keep this issue open as a reminder.

You can actually improve your use case a lot using this:

```js
// Stryker disable all: Don't mutate compiler macros
defineProps({
  msg: { type: String, required: true },
  anotherProp: { type: String, required: false }
})
// Stryker restore all
```

See https://stryker-mutator.io/docs/stryker-js/disable-mutants/#using-a--stryker-disable-comment for the docs

**Comment by nicojs** [comment]
Thanks, but we want to keep this open. We shouldn't ignore the ignore plugin use case

**Comment by davidlandais** [comment]
Hello @nicojs,

I ran into the same issue on a large Nuxt codebase. I wanted to add StrykerJS as an extra layer of confidence for my own code and to make sure what comes out of AI-assisted coding holds up to real mutation testing.

The usual workaround of adding `// Stryker disable all` and `// Stryker restore all` around every macro in every .vue file. Is not really viable with a codebase with 100+ components. It adds noise to every single SFC and it's easy to forget on new components.

This issue was opened in 2021 and closed by the stale bot in 2025 without a fix. We're in 2026 now. Angular has had a built-in ignorer since v8.3, but Vue still has nothing.

So I coded a small plugin that does the job. It follows the same Ignorer pattern as the built-in Angular ignorer, but the scope is a bit different: Vue macros are compile-time directives that get entirely extracted by the compiler, so the plugin ignores the whole CallExpression rather than just an ObjectExpression argument.

It works on my end, tested on ~100 .vue components. The repo is here: https://github.com/davidlandais/stryker-vue-ignorer

Feel free to use it. I think this deserves a built-in solution rather than staying an unresolved issue closed by a bot.
